VMTouch.cpp

```diff
@@ -0,0 +1,184 @@
+    int res;
+    int open_flags;
+
+    open_flags = O_RDONLY;
+    open_flags |= O_NOATIME;
+
+    try {
+        fd = open(path, open_flags, 0);
+
+        if (fd == -1 && errno == EPERM) {
+            open_flags &= ~O_NOATIME;
+            fd = open(path, open_flags, 0);
+        }
+
+        if (fd == -1) {
+            SWARN("unable to open " << path << " (" << strerror(errno) << "), skipping");
+            throw SException("Unable to open");
+        }
+
+        res = fstat(fd, &sb);
+        if (res) {
+            SWARN("unable to fstat " << path << " (" << strerror(errno) << "), skipping");
+            throw SException("unable to fstat");
+        }
+
+        if (S_ISBLK(sb.st_mode)) {
+            if (ioctl(fd, BLKGETSIZE64, &len_of_file)) {
+                SWARN("unable to ioctl " << path << " (" << strerror(errno) << "), skipping");
+                throw SException("unable to ioctl");
+            }
+        } else {
+            len_of_file = sb.st_size;
+        }
+
+        if (len_of_file == 0) {
+            throw SException("Empty file");
+        }
+
+        mem = mmap(NULL, len_of_file, PROT_READ, MAP_SHARED, fd, 0);
+
+        if (mem == MAP_FAILED) {
+            SWARN("unable to mmap file " << path << " (" << strerror(errno) << "), skipping");
+            throw SException("unable to mmap");
+        }
+
+        if (!aligned_p(mem)) {
+            SERROR("mmap(" << path << ") wasn't page aligned");
+        }
+
+        pages_in_range = bytes2pages(len_of_file);
+
+        char* mincore_array = (char*)malloc(pages_in_range);
+        if (mincore_array == NULL) {
+            SWARN("Failed to allocate memory for mincore array (" << strerror(errno) << ")");
+        }
+
+        if (mincore(mem, len_of_file, (unsigned char*)mincore_array)) {
+            SERROR("mincore " << path << " (" << strerror(errno) << ")");
+        }
+
+        if (o_touch) {
+            if (verbose) {
+                cout << "Touching memory." << endl;
+            }
+            list<thread> counters;
+            const size_t thread_count = thread::hardware_concurrency();
+            int64_t pages_per_thread = (pages_in_range / thread_count) + 1;
+            for (size_t i = 0; i < thread_count; i++) {
+                counters.emplace_back([i, &pages_per_thread, &pages_in_range, &mincore_array, &mem]() {
+                    unsigned int junk_counter = 0;
+                    int64_t first_page = pages_per_thread * i;
+                    int64_t last_page = min(first_page + pages_per_thread, pages_in_range - 1);
+                    for (int64_t j = first_page; j < last_page; j++) {
+                        if (!is_mincore_page_resident(mincore_array[j])) {
+                            junk_counter += ((char*)mem)[j * pagesize];
+                        }
+                    }
+
+                    // Passing junk_counter here avoids a "set but not used" error.
+                    do_nothing(junk_counter);
```

Comment:
Could we instead remove the `junk_count +=` part of line 121 and remove this or would that make it be completely removed by the compiler?

( I do like the fact that we have a function called do_nothing 😂 )

Reply:
I haven't investigated what the compiler will do with this, I just kept this from the original implementation.

If we do nothing whatsoever with this value, the compiler will probably optimize out the read, though.